### DISCUSSION TITLE
How to highlight multiple lines and make clickable line number.

### DISCUSSION THREAD

**Discussion by ViStefan** [ajaxorg/ace#5194] [Q&A]
Hello. I want to use Ace in code editor window with debugging functionality. So I want to make editor to highlight current executing line despite what is current cursor position. I also need to highlight multiple lines (e.g. all the calls in stacktrace). Ace documentation is not very easy to dive in for me. Sorry if I'm asking something obvious, but my question is:
What are (and if there are) API methods in Ace to:
1. highlight the line;
2. highlight multiple lines; and
3. make clickable line numbers in editor (to toggle breakpoints).

Thank you.

**Comment by nightwing** [comment]
for callstack you can use session.addMarker method like https://github.com/c9/core/blob/master/plugins/c9.ide.run.debug/callstack.js#L363
ace has builtin styles for ace_step and ace_stack, but you can also use other names and add custom css for the marker
```
var Range = ace.Range 
editor.session.addMarker(new Range(0,0,0,1), "ace_step", "fullLine")
editor.session.addMarker(new Range(3,0,3,1), "ace_stack", "fullLine")
```

for breakpoints there is `editor.session.setBreakpoint(row, className)` method, which adds classname to gutter element. The code using that in cloud9 is at https://github.com/c9/core/blob/master/plugins/c9.ide.run.debug/breakpoints.js
